### PR TITLE
[sBTC DR] Withdrawal flow

### PR DESCRIPTION
**Pull request by stjepangolemac**
- Parses withdrawal requests from the bitcoin state
- Burns sBTC
- Withdraws BTC to the destination address
- It now separately tracks Bitcoin and Stacks blocks
- Requires RPC Bitcoin node access
- Requires Electrum service access

### CODE REVIEW

File: romeo/asset-contract/contracts/clarity-bitcoin-mini.clar
Comment (radicleart) [inline romeo/asset-contract/contracts/clarity-bitcoin-mini.clar:1]: There's an issue with this change - verify-block-header should be called and DEBUG_MODE needs to be set to false on testnet deployment.

**Review by radicleart — COMMENTED** [review]
on deploying the contracts to testnet it came out that DEBU_MODE In calrity-bitcoin-mini needs to be switched off and `was-tx-mined` needs to call `verify-block-header`

File: stacks-core/src/address.rs [hunk 1]
```diff
@@ -35,6 +32,40 @@ pub enum AddressVersion {
...
+/// Supported stacks address versions
+#[repr(u8)]
+#[derive(FromRepr, EnumIter, EnumString, PartialEq, Eq, Copy, Clone, Debug)]
+#[strum(ascii_case_insensitive)]
+pub enum AddressKind {
+    /// Pay to public key hash
+    P2PKH = 0,
+    /// Pay to script hash
+    P2SH = 1,
+    /// Pay to witness public key hash
```
Comment (friedger) [inline stacks-core/src/address.rs:44]: Maybe add a comment that this is the only AddressKind used in Stacks for single accounts

**Review by friedger — CHANGES_REQUESTED** [review]
Looks good apart from the change to clarity-bitcoin-mini

File: romeo/asset-contract/contracts/clarity-bitcoin-mini.clar
Comment (stjepangolemac) [reply romeo/asset-contract/contracts/clarity-bitcoin-mini.clar:1]: Okay I guess it makes sense for debug mode to be disabled by default.

File: stacks-core/src/address.rs (hunk 1 above)
Comment (stjepangolemac) [reply stacks-core/src/address.rs:44]: It's actually a p2pkh if you look at the `Address::from_public_key` function. I'm not sure at the moment which one should it be.

**Comment by AshtonStephens** [comment]
For my own curiosity - is there a diagram of this flow anywhere?

File: stacks-core/src/address.rs (hunk 1 above)
Comment (friedger) [reply stacks-core/src/address.rs:44]: p2pkh looks correct

**Comment by stjepangolemac** [comment]
> For my own curiosity - is there a diagram of this flow anywhere?

No, is that something we'd like to have in the docs or only here to make the review clearer?

File: stacks-core/src/address.rs (hunk 1 above)
Comment (stjepangolemac) [reply stacks-core/src/address.rs]: I removed AddressKind completely. It's probably a remnant of the code I used to signal what kind of Stacks address should be used to burn from. We now by convention always use P2PKH for that.

File: .gitignore [hunk 2]
```diff
@@ -4,4 +4,4 @@ coverage
 romeo/testing/
 romeo/asset-contract/.test
 romeo/asset-contract/.coverage
-**/.DS_Store
+**/romeo.log
```
Comment (friedger) [inline .gitignore:7]: DS_Store should stay

File: romeo/asset-contract/contracts/clarity-bitcoin-mini.clar [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 ;; it can only check if a txid is part of a burn chain block
 
-(define-constant DEBUG-MODE true)
+(define-constant DEBUG-MODE false)
```
Comment (friedger) [inline romeo/asset-contract/contracts/clarity-bitcoin-mini.clar:5]: Now unit tests fail

**Review by friedger — CHANGES_REQUESTED** [review]

File: romeo/asset-contract/contracts/clarity-bitcoin-mini.clar (hunk 3 above)
Comment (stjepangolemac) [reply romeo/asset-contract/contracts/clarity-bitcoin-mini.clar]: I'll revert the debug mode change as that's being fixed in a separate PR.

**Review by friedger — APPROVED** [review]
All good! Great work!

**Comment by AshtonStephens** [comment]
Lots of code cov warnings here - is none of this actually covered? Also - how are we sure that this works? What local testing did you run so I can recreate it?

File: devenv/utils/deposit.sh [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+json=$(sbtc deposit \
+    -w $btc_wif \
+    -n regtest \
+    -r $stacks_address \
+    -a $amount \
+    -d $btc_p2tr_address \
+    -u localhost:60401)
+
+tx=$(echo -n $json | jq -r .hex)
```
Comment (friedger) [inline devenv/utils/deposit.sh:19]: Can we show some output if the deposit fails?

File: devenv/utils/mine_btc.sh [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+num_blocks=$1
+dir="$(dirname "$0")"
+
+if [[ -z "$num_blocks" ]]; then
+    echo "Usage: ./script.sh [num_blocks]"
+    exit 1
+fi
+
+btc_address=$(source $dir/get_credentials.sh | jq -r '.credentials["0"].bitcoin.p2wpkh.address')
```
Comment (friedger) [inline devenv/utils/mine_btc.sh:11]: We need also bitcoins for p2pkh:

btc_address=$(source $dir/get_credentials.sh | jq -r '.credentials["0"].bitcoin.p2pkh.address')

**Comment by AshtonStephens** [comment]
@stjepangolemac since you're out next week it may make the most sense to merge this now and then iterate on any bugs that we catch with patches to the main branch.

**Review by AshtonStephens — APPROVED** [review]
Approving so this is merged knowing it lacks coverage and isn't perfect.

File: romeo/src/bitcoin_client.rs [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+			return Err(anyhow::anyhow!("Password is empty"));
+		}
+
+		url.set_username("").unwrap();
+		url.set_password(None).unwrap();
+
+		let client =
+			RPCClient::new(url.as_ref(), Auth::UserPass(username, password))?;
+
+		Ok(spawn_blocking(move || f(client)).await?)
```
[the remark this answers is not in the document]
Comment (stjepangolemac) [reply romeo/src/bitcoin_client.rs:61]: The underlying client implementation is blocking and by spawning this to a separate blocking worker thread we're not blocking the runtime.

File: devenv/utils/deposit.sh [hunk 7]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (stjepangolemac) [reply devenv/utils/deposit.sh:1]: Yep, sure.

File: devenv/utils/deposit.sh (hunk 4 above)
Comment (stjepangolemac) [reply devenv/utils/deposit.sh:19]: Yeah can try to do something like that.

Comment (stjepangolemac) [reply devenv/utils/deposit.sh:19]: Added this to https://github.com/stacks-network/sbtc/pull/175

File: devenv/utils/mine_btc.sh (hunk 5 above)
Comment (stjepangolemac) [reply devenv/utils/mine_btc.sh:11]: Added this to https://github.com/stacks-network/sbtc/pull/175

File: devenv/utils/deposit.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (stjepangolemac) [reply devenv/utils/deposit.sh:1]: Added this to https://github.com/stacks-network/sbtc/pull/175

File: romeo/src/bitcoin_client.rs [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+#[derive(Debug, Clone)]
+pub struct Client(Config);
+
+impl Client {
+	/// Create a new RPC client
+	pub fn new(config: Config) -> anyhow::Result<Self> {
+		Ok(Self(config))
+	}
+
+	async fn execute<F, T>(
```
[the remark this answers is not in the document]
Comment (stjepangolemac) [reply romeo/src/bitcoin_client.rs:34]: I can add some docs but will add them in https://github.com/stacks-network/sbtc/pull/175

**Pull request merged**